File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/providers/portfolio_status.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass(frozen=True)
class ProviderResult:
    values: dict[str, str]
    text: str
    data: dict[str, str | int | float]


class PortfolioStatusProvider:
    @property
    def name(self) -> str:
        return "PORTFOLIO_STATUS"

    @property
    def description(self) -> str:
        return "Provides current portfolio holdings, PnL, and trading state"

    @property
    def position(self) -> int:
        return 50

    async def get(
        self,
        _message: Message,
        _state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ProviderResult:
        if service is None:
            return ProviderResult(
                values={
                    "portfolioStatus": "Trading service is not available",
                    "tradingState": "unknown",
                },
                text="# Portfolio Status\n\nTrading service is not available.",
                data={"holdings": 0, "state": "unknown"},
            )

        portfolio = await service.check_portfolio()
        state = await service.get_state()
        config = await service.get_strategy_config()
        history = await service.get_trade_history(5)

        text_parts = [
            "# Portfolio Status",
            "",
            f"State: {state}",
            f"Strategy: {config.strategy.value}",
            f"Total Value: ${portfolio.total_value:.2f} "
            f"| PnL: ${portfolio.pnl:.2f} ({portfolio.pnl_pct:+.2f}%)",
            "",
            "## Holdings",
        ]

        if not portfolio.holdings:
            text_parts.append("No open positions.")
        else:
            for token, h in portfolio.holdings.items():
                text_parts.append(
                    f"- {token}: {h.amount:.4f} @ ${h.current_price:.4f} "
                    f"= ${h.value:.2f} (PnL: ${h.pnl:.2f})"
                )

        if history:
            text_parts.append("")
            text_parts.append("## Recent Trades")
            for t in history:
                text_parts.append(
                    f"- {t.timestamp.strftime('%H:%M:%S')} {t.direction} "
                    f"{t.amount:.4f} {t.token} @ ${t.price:.4f}"
                )

        return ProviderResult(
            values={
                "portfolioStatus": (
                    f"${portfolio.total_value:.2f} "
                    f"| PnL: ${portfolio.pnl:.2f} ({portfolio.pnl_pct:+.2f}%)"
                ),
                "tradingState": state,
                "strategy": config.strategy.value,
            },
            text="\n".join(text_parts),
            data={
                "holdings": len(portfolio.holdings),
                "totalValue": portfolio.total_value,
                "pnl": portfolio.pnl,
                "state": state,
                "recentTrades": len(history),
            },
        )

File: plugins/plugin-auto-trader/rust/src/providers/portfolio_status.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::{Provider, ProviderResult};

pub struct PortfolioStatusProvider;

#[async_trait]
impl Provider for PortfolioStatusProvider {
    fn name(&self) -> &str {
        "PORTFOLIO_STATUS"
    }

    fn description(&self) -> &str {
        "Provides current portfolio holdings, PnL, and trading state"
    }

    fn position(&self) -> i32 {
        50
    }

    async fn get(
        &self,
        _message: &Value,
        _state: &Value,
        service: Option<&TradingService>,
    ) -> ProviderResult {
        let Some(svc) = service else {
            return ProviderResult {
                values: json!({
                    "portfolioStatus": "Trading service is not available",
                    "tradingState": "unknown"
                }),
                text: "# Portfolio Status\n\nTrading service is not available.".to_string(),
                data: json!({ "holdings": 0, "state": "unknown" }),
            };
        };

        let portfolio = svc.check_portfolio().await;
        let state = svc.get_state().await;
        let config = svc.get_strategy_config().await;
        let history = svc.get_trade_history(5).await;

        let state_str = format!("{:?}", state);

        let mut text_parts = vec![
            format!("# Portfolio Status"),
            format!(""),
            format!("State: {}", state_str),
            format!("Strategy: {}", config.strategy),
            format!(
                "Total Value: ${:.2} | PnL: ${:.2} ({:+.2}%)",
                portfolio.total_value, portfolio.pnl, portfolio.pnl_pct
            ),
            format!(""),
            format!("## Holdings"),
        ];

        if portfolio.holdings.is_empty() {
            text_parts.push("No open positions.".to_string());
        } else {
            for (token, h) in &portfolio.holdings {
                text_parts.push(format!(
                    "- {}: {:.4} @ ${:.4} = ${:.2} (PnL: ${:.2})",
                    token, h.amount, h.current_price, h.value, h.pnl
                ));
            }
        }

        if !history.is_empty() {
            text_parts.push(String::new());
            text_parts.push("## Recent Trades".to_string());
            for t in &history {
                text_parts.push(format!(
                    "- {} {} {:.4} {} @ ${:.4}",
                    t.timestamp.format("%H:%M:%S"),
                    t.direction,
                    t.amount,
                    t.token,
                    t.price
                ));
            }
        }

        ProviderResult {
            values: json!({
                "portfolioStatus": format!(
                    "${:.2} | PnL: ${:.2} ({:+.2}%)",
                    portfolio.total_value, portfolio.pnl, portfolio.pnl_pct
                ),
                "tradingState": state_str,
                "strategy": config.strategy.to_string(),
            }),
            text: text_parts.join("\n"),
            data: json!({
                "holdings": portfolio.holdings.len(),
                "totalValue": portfolio.total_value,
                "pnl": portfolio.pnl,
                "state": state_str,
                "recentTrades": history.len(),
            }),
        }
    }
}
